drv/spartan7-loader/src/main.rs

```diff
@@ -0,0 +1,230 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Dedicated task for loading the Spartan-7 FPGA bitstream
+//!
+//! This FPGA is used as a memory-mapped peripheral.  Once it's loaded, other
+//! tasks interact with it by writing directly to memory addresses, so it's
+//! important that it remains running without interruption.  As such, this task
+//! is infallible once it enters the Idol runtime loop.
+
+#![no_std]
+#![no_main]
+
+use core::num::NonZeroUsize;
+use drv_spartan7_spi_program::{BitstreamLoader, Spartan7Error};
+use drv_spi_api::{SpiDevice, SpiServer};
+use drv_stm32xx_sys_api as sys_api;
+use idol_runtime::{NotificationHandler, RequestError};
+use userlib::{hl, sys_recv_notification, task_slot, RecvMessage, UnwrapLite};
+
+use ringbuf::{counted_ringbuf, ringbuf_entry, Count};
+
+////////////////////////////////////////////////////////////////////////////////
+// Select local vs server SPI communication
+
+/// Claims the SPI core.
+///
+/// This function can only be called once, and will panic otherwise!
+#[cfg(feature = "use-spi-core")]
+pub fn claim_spi(
+    sys: &sys_api::Sys,
+) -> drv_stm32h7_spi_server_core::SpiServerCore {
+    drv_stm32h7_spi_server_core::declare_spi_core!(
+        sys.clone(),
+        notifications::SPI_IRQ_MASK
+    )
+}
+
+#[cfg(not(feature = "use-spi-core"))]
+pub fn claim_spi(_sys: &sys_api::Sys) -> drv_spi_api::Spi {
+    task_slot!(SPI, spi);
+    drv_spi_api::Spi::from(SPI.get_task_id())
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Copy, Clone, PartialEq, Count)]
+enum Trace {
+    FpgaInit,
+    FpgaInitFailed(#[count(children)] Spartan7Error),
+    StartFailed(#[count(children)] SeqError),
+    ContinueBitstreamLoad(usize),
+    WaitForDone,
+    Programmed,
+
+    #[count(skip)]
+    None,
+}
+
+#[derive(Copy, Clone, PartialEq, Count)]
+enum SeqError {
```

Comment:
nitpick, feel free to ignore me: I know this was just copied from `grapefruit-sesq-server`, but...does it really make sense to call this `SeqError` now that it's in the `spartan7-loader` driver? Should it instead be called `LoaderError` or something now?

Reply:
Sure, done in 73b9724b